googledest/sheets.go

```diff
@@ -0,0 +1,210 @@
+package googledest
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sheets/v4"
+
+	sync "github.com/silinternational/personnel-sync/v3"
+)
+
+const DefaultSheetName = "Sheet1"
+
+type GoogleSheets struct {
+	DestinationConfig sync.DestinationConfig
+	GoogleConfig      GoogleConfig
+	Service           *sheets.Service
+	SheetsSyncSet     SheetsSyncSet
+}
+
+type SheetsSyncSet struct {
+	SheetID   string
+	SheetName string
+}
+
+func NewGoogleSheetsDestination(destinationConfig sync.DestinationConfig) (sync.Destination, error) {
+	var s GoogleSheets
+
+	err := json.Unmarshal(destinationConfig.ExtraJSON, &s.GoogleConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling GoogleConfig: %s", err)
+	}
+
+	s.Service, err = initSheetsService(
+		s.GoogleConfig.GoogleAuth,
+		s.GoogleConfig.DelegatedAdminEmail,
+		sheets.SpreadsheetsScope,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing Google Sheets service: %s", err)
+	}
+
+	s.DestinationConfig = destinationConfig
+
+	return &s, nil
+}
+
+func initSheetsService(auth GoogleAuth, adminEmail string, scopes ...string) (*sheets.Service, error) {
+	googleAuthJson, err := json.Marshal(auth)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal google auth data into json, error: %s", err.Error())
+	}
+
+	config, err := google.JWTConfigFromJSON(googleAuthJson, scopes...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse client secret file to config, error: %s", err)
+	}
+
+	config.Subject = adminEmail
+
+	ctx := context.Background()
+	svc, err := sheets.NewService(ctx, option.WithHTTPClient(config.Client(ctx)))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create sheets service, error: %s", err)
+	}
+	return svc, nil
+}
+
+func (g *GoogleSheets) GetIDField() string {
+	return ""
+}
+
+func (g *GoogleSheets) ForSet(syncSetJson json.RawMessage) error {
+	var syncSetConfig SheetsSyncSet
+	err := json.Unmarshal(syncSetJson, &syncSetConfig)
+	if err != nil {
+		return err
+	}
+
+	g.SheetsSyncSet = syncSetConfig
+
+	// Defaults
+	if g.SheetsSyncSet.SheetName == "" {
+		g.SheetsSyncSet.SheetName = DefaultSheetName
+	}
+
+	return nil
+}
+
+func (g *GoogleSheets) ListUsers() ([]sync.Person, error) {
+	var members []sync.Person
+
+	// To keep it simple, ignore the existing content and overwrite the entire sheet
+
+	return members, nil
+}
+
+func (g *GoogleSheets) ApplyChangeSet(
+	changes sync.ChangeSet,
+	eventLog chan<- sync.EventLogItem) sync.ChangeResults {
+
+	if g.DestinationConfig.DisableAdd || g.DestinationConfig.DisableDelete || g.DestinationConfig.DisableUpdate {
```

Comment:
Looking at this, I'm a little confused why we have all three configs when only one of them is needed.

Reply:
It's defined at the `Destination` level of the config, so any adapter can use it. Another option would be to add a new config option `Disable` but that's nearly redundant with `DryRunMode`.

Reply:
:+1: